cmd/oras/root/cp.go

```diff
@@ -191,3 +195,40 @@ func runCopy(ctx context.Context, opts copyOptions) error {
 
 	return nil
 }
+
+// RecursiveCopy copies an artifact and its referrers from one target to another.
+// If the artifact is a manifest list or index, referrers of its manifests are copied as well.
+func RecursiveCopy(ctx context.Context, src oras.ReadOnlyGraphTarget, dst oras.Target, dstRef string, root ocispec.Descriptor, opts oras.ExtendedCopyOptions) error {
+	var err error
+	if root.MediaType == ocispec.MediaTypeImageIndex || root.MediaType == docker.MediaTypeManifestList {
+		var fetched []byte
+		fetched, err := content.FetchAll(ctx, src, root)
+		if err != nil {
+			return err
+		}
+		var index ocispec.Index
+		if err = json.Unmarshal(fetched, &index); err != nil {
+			return nil
+		}
+
+		var wg sync.WaitGroup
+		for _, desc := range index.Manifests {
+			wg.Add(1)
+			go func(desc ocispec.Descriptor) {
+				defer wg.Done()
+				err = oras.ExtendedCopyGraph(ctx, src, dst, desc, opts.ExtendedCopyGraphOptions)
+			}(desc)
+		}
+		wg.Wait()
+		if err != nil {
+			return err
+		}
```

Comment:
Should you copy the index/manifest list here and return, instead of going into the next `if` and then `ExtendedCopyGraph`/`ExtendedCopy`?

Reply:
It's done in the below if-else block.

Reply:
I see that we don't have a public `copyNode` function in `oras-go`, so probably we can only do it this way.

Reply:
> Should you copy the index/manifest list here and return

Still the index/manifest list should be copied with its referrers.